server/modules/formula.py

```diff
@@ -14,29 +25,85 @@ def render(wf_module, table):
         if formula == '':
             return table    # nop if no formula
 
-        colnames = [x.replace(" ", "_") for x in table.columns]  # spaces to underscores in column names
-        newcol = pd.Series(np.zeros(len(table)))
+        syntax = wf_module.get_param_menu_idx('syntax')
+
+        newcol = pd.Series(np.zeros(len(table)), dtype=np.dtype(object))
+
+        if syntax == 0:  # Python
+            colnames = [x.replace(" ", "_") for x in table.columns]  # spaces to underscores in column names
+
+            # Catch errors with the formula and display to user
+            try:
+                code = compile(formula, '<string>', 'eval')
+
+                # Much experimentation went into the form of this loop for good performance.
+                # Note we don't use iterrows or any pandas indexing, and construct the values dict ourselves
+                for i,row in enumerate(table.values):
+                    newcol[i] = eval(code, custom_code_globals, dict(zip(colnames, row)))
+            except Exception as e:
+                return(str(e))
+
+        if syntax == 1:  # Excel
+            try:
+                code = Parser().ast(formula)[1].compile()
+            except Exception as e:
+                return "Couldn't parse formula: %s" % str(e)
```

Comment:
All new strings need to be localizable by passing them to the _ function. See e.g. https://github.com/CJWorkbench/cjworkbench/blob/master/server/dispatch.py#L165